Documentation/app-operations.md

```diff
@@ -0,0 +1,171 @@
+
+The motivation behind this change is the direction orchestration systems are taking
+and to facilitate this new direction. For more details, see [kubernetes#25899](https://github.com/yujuhong/kubernetes/blob/08dc66113399c89e31f6872f3c638695a6ec6a8d/docs/proposals/container-runtime-interface-v1.md).
+
+# API
+
+The envisioned workflow for the app-level CLI API is that After a pod has been started, 
+the users will be invoking the rkt CLI which is application logic on top of stage1 entrypoints (introduced below).
+
+## rkt app inject
+Injects an application image into a running pod.
+
+```bash
+rkt app inject <pod-uuid> <image-name/hash/address/registry-URL> <arguments>
+```
+
+## rkt app eject
+Ejects an application image into a running pod, meaning that the traces of it are removed from the pod.
+
+```bash
+rkt app eject <pod-uuid> --app=<app-name> <arguments>
+```
+
+**Note:** Currently we don't run garbage collection while the pod is still running, should
+we consider doing it once we introduce ejection?
+
+## rkt app start
+Starts an application that was either stopped or injected.
+This operation is idempotent.
+
+```bash
+rkt app start <pod-uuid> --app=<app-name> <arguments>
+```
+
+## rkt app stop
+Stops a running application gracefully.
+
+```bash
+rkt app stop <pod-uuid> --app=<app-name> <arguments>
+```
+
+## rkt app list
+Lists the applications that are inside a pod, running or stopped.
+
+```bash
+rkt app list <pod-uuid> <arguments>
+```
+
+## rkt app status
+Returns the execution status of application inside a pod.
+
+```bash
+rkt app status <pod-uuid> --app=<app-name> <arguments>
+```
+
+## rkt app exec
+Executes a command inside an application.
+
+```bash
+rkt app exec <pod-uuid> --app=<app-name> <arguments> -- <command> <command-arguments>
+```
+
+# Entrypoints
+
+In order to facilitate the app-level operations API, 4 new entrypoints are introduced.
+These entrypoints are stage1 binaries that are capable of communicating with the stage1 supervisor.
+
+For example, the `app start` entrypoint in default stage1 flavor sends a `start` signal
+(similar to `systemctl start`) for the service files of the application to the systemd which is the stage1 supervisor.
+
+## rkt app inject
+
+`coreos.com/rkt/stage1/app/inject`
+
+1. prepare the application intended for injection, its rootfs and service files or else.
+2. `coreos.com/rkt/stage1/inject` entrypoint is resolved via annotations found within `/var/lib/rkt/pods/run/$uuid/stage1/manifest`.
+3. the entrypoint will be executed and passed a reference to the rootfs of the application that was prepared.
+
+The responsibility of this entrypoint is to receive a prepared app, copy its contents into stage2, 
+so that it is ready to be started by the stage1 supervisor (e.g. systemd for the default flavor).
+
+For example the operations would be as follows for the default rkt stage1 flavor:
+- Prepare the application rootfs and its service files given an application image/images.
```

Comment:
This steps is done by different stages:

* stage0: Prepare the application rootfs: it can be: direct rendering or overlayfs over the treestore
* stage1: Prepare the application service files given an application image/images.